index.js

```diff
@@ -1,22 +1,60 @@
-module.exports = Hook
-
 var register = require('./lib/register')
 var addHook = require('./lib/add')
 var removeHook = require('./lib/remove')
 
-function Hook () {
+var bind = Function.bind
+var bindable = bind.bind(bind)
+
+function bindApi (hook, state, name) {
+  const removeHookRef = bindable(removeHook, null).apply(null, name ? [state, null, name] : [state, null])
+  hook.api = { remove: removeHookRef }
+  hook.remove = removeHookRef
+
+  ;['before', 'error', 'after', 'wrap'].forEach(function (kind) {
+    const args = name ? [state, kind, name] : [state, kind]
+    hook[kind] = hook.api[kind] = bindable(addHook, null).apply(null, args)
+    hook.remove[kind] = hook.api.remove[kind] = bindable(removeHook, null).apply(null, args)
+  })
+}
+
+var singularHookState = {
+  registry: {}
+}
```

Comment:
🤔 Having a state in global is a memory leak ... I feel like we’d need to change the API to work around that ... any thoughts?

We could solve it by going back to your initial suggestion of adding a method to the current API. Maybe that’s why you implemented it this way in the first place? Hmmm I’m so sorry for the back and forth, not sure what best to do, I’ll give it some more thought

Reply:
Nah, it's ok, I like this back and forth because I didn't even realise this is a memory leak 😉 
But can you elaborate why this would be a memory leak? Doesn't the remove method clear up the state when called?
As you can see, I did do any of the things I initialy did for a purpose like this ;)

But I guess we can wrap it inside another method if you think that's necessary.

Reply:
> Doesn't the remove method clear up the state when called?

If you explicitly call the `.remove()` then it works. But say someone creates a lot of these singular hooks, for example in a request handler. As a developer I would assume that all references are removed by the garbage collection once the request handler is done. But in this case the GC cannot clear up `singularHookState` because it is used statically. 

The Hook Collection API on the other hand sets its state in the constructor. When the hook collection instance is no longer referenced anywhere in the code, it can be cleaned up entirely by GC

Reply:
Ah, I see, that makes sense. I've moved the state inside the SingularHook constructor. It uses the Collection logic but it will always be limited to just one hook. It works and still makes sense, I like it 😄